text/0753-logging-and-debugging-context-api.md

```diff
@@ -0,0 +1,189 @@
+through its parents on demand, building up the debugging context as we do.
+
+Since we are utilizing an existing datastructure which _necessarily_ exists in
+production apps, we won't be introducing any additional costs by adding this
+API. We also won't need to retrofit this new API into the VM, as the
+datastructure already exists. In addition, because users can add onto the
+destroyable tree themselves, this API can also be used to provide context for
+user constructs, making it extensible.
+
+## Detailed design
+
+As discussed in the motivation section, this API will build on the destroyables
+API, using it as the basis for how it operates. We will introduce the following
+new functions:
+
+```ts
+interface PartialContextDescriptor {
+  description: string;
+  extra?: unknown;
+}
+
+interface ContextDescriptor extends PartialContextDescriptor {
+  parent: ContextDescriptor | ContextDescriptor[] | null;
+  isDestroying: boolean;
+  isDestroyed: boolean;
+}
+
+declare function getDebuggingContext(obj: object): ContextDescriptor | null;
+declare function setDebuggingContextGenerator(
+  obj: object,
+  generator: (obj: object) => PartialContextDescriptor
+): void;
+```
+
+These functions will be importable from the `@glimmer/debug` module, since they
+have to do with debugging. Unlike other debugging functions, like the ones
+provided in `@ember/debug`, these will _not_ be removed from applications in
+production builds since they will still provide useful behavior, and will not
+incur performance penalties until they are used.
+
+### getDebuggingContext
+
+`getDebuggingContext` is passed an object, and returns either a descriptor of
+that object which can be used for debugging purposes, or `null` if the object
+does not have any debugging context generators associated with it or any of its
+parents. The debugging context object contains the following properties:
+
+- `description`: A string which describes the object itself
+- `parent`: If the object has a parent (see below) which also has a debugging
+  context, then this property is that context object. If there are multiple
+  such parents, then it is an array of objects. Otherwise, it is `null`.
+
+Debugging context parents are the same parents that are associated with the
+object via `associateDestroyableChild`. If a particular parent exists, but does
+not have a debugging context generator, then it is skipped and its parent(s) are
+used instead.
+
+Debugging context generators for a given object are set via
+`setDebuggingContextGenerator`. When you call `getDebuggingContext` on a
+particular object, it first checks to see if a generator was associated directly
+with that object. If one wasn't, it then checks the prototypes of that object.
+This way, debugging context generators can be associated with classes instead of
+individual objects. When one is found, it is called and passed the object that
+`getDebuggingContext` was called on.
+
+### setDebuggingContextGenerator
```

Comment:
Can you update to include information about what happens if `setDebuggingContextGenerator` is called twice?